services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRequestObjectExecutor.java

```diff
@@ -0,0 +1,508 @@
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.fasterxml.jackson.databind.JsonNode;
+
+public class SecureRequestObjectExecutor implements ClientPolicyExecutorProvider {
+
+    private static final Logger logger = Logger.getLogger(SecureRequestObjectExecutor.class);
+
+    private final KeycloakSession session;
+    private final ComponentModel componentModel;
+
+    private AuthorizationEndpointRequestObject parsedRequestObject;
+
+    public static final String INVALID_REQUEST_OBJECT = "invalid_request_object";
+
+    public SecureRequestObjectExecutor(KeycloakSession session, ComponentModel componentModel) {
+        this.session = session;
+        this.componentModel = componentModel;
+    }
+
+    @Override
+    public void executeOnEvent(ClientPolicyContext context) throws ClientPolicyException {
+        switch (context.getEvent()) {
+            case AUTHORIZATION_REQUEST:
+                AuthorizationRequestContext authorizationRequestContext = (AuthorizationRequestContext)context;
+                executeOnAuthorizationRequest(authorizationRequestContext.getparsedResponseType(),
+                    authorizationRequestContext.getAuthorizationEndpointRequest(),
+                    authorizationRequestContext.getRedirectUri());
+                break;
+            default:
+                return;
+        }
+    }
+
+    public void executeOnAuthorizationRequest(
+            OIDCResponseType parsedResponseType,
+            AuthorizationEndpointRequest request,
+            String redirectUri) throws ClientPolicyException {
+        ClientPolicyLogger.log(logger, "Authz Endpoint - authz request");
+
+        // Only support GET
+        MultivaluedMap<String, String> params = session.getContext().getUri().getQueryParameters();
+        String requestParam = params.getFirst(OIDCLoginProtocol.REQUEST_PARAM);
+        String requestUriParam = params.getFirst(OIDCLoginProtocol.REQUEST_URI_PARAM);
+
+        // check whether whether request object exists
+        if (requestParam == null && requestUriParam == null) {
+            ClientPolicyLogger.log(logger, "request object not exist.");
+            throw new ClientPolicyException(OAuthErrorException.INVALID_REQUEST, "Invalid parameter");
+        }
+
+        // check whether request_uri is https scheme
+        if (requestUriParam != null && !requestUriParam.startsWith("https://")) {
+            ClientPolicyLogger.log(logger, "request_uri scheme is not https.");
+            throw new ClientPolicyException(OAuthErrorException.INVALID_REQUEST, "Invalid parameter: request_uri");
+        }
+
+        // check whether request object can be retrieved from request_uri
+        String retrievedRequestObject = null;
+        if (requestParam != null) {
+            retrievedRequestObject = requestParam;
+        } else {
+            try (InputStream is = session.getProvider(HttpClientProvider.class).get(requestUriParam)) {
```

Comment:
This means that when "request_uri" is used, then there will be 2 HTTP requests needed to "request_uri" during single AuthorizationEndpoint request. Because it is also loaded and parsed in the AuthzEndpointRequestObjectParser.

Also even if "request" object is used (and not "request_uri"), the parsing of "request" object is duplicated currently etc.

I hope it is possible to somehow share the loading/parsing of the "request" object with the AuthzEndpointRequestObjectParser? This would mean that it would need to be done only once per AuthorizationEndpoint request and also will avoid some code duplications. Maybe parsed "request" can be saved as attribute of KeycloakSession for example?

Reply:
@mposolda To avoid loading and parsing request object, how about the following ?

- `AuthzEndpointRequestObjectParser` do verification that currently `SecureRequestObjectExecutor` do and storing its result
- `SecureRequestObjectExecutor` check this stored result and throw an appropriate `ClientPolicyException`.

Reply:
Also, I'll investigate the way you suggested, namely parsed "request" being saved as attribute of KeycloakSession.

Reply:
@tnorimat It seems to me that some basic verifications like those lines https://github.com/Hitachi/keycloak/blob/KEYCLOAK-14204/services/src/main/java/org/keycloak/services/clientpolicy/executor/SecureRequestObjectExecutor.java#L90-L105 can be still kept on `SecureRequestObjectExecutor`. However the downloading from "request_uri" and parsing "request" can be done on `AuthzEndpointRequestObjectParser` and then saved as attribute in the KeycloakSession. If there is some issue in downloading the "request_uri" or parsing the "request" object, then the exception will be thrown earlier in the `AuthzEndpointRequestObjectParser` (not in the `SecureRequestObjectExecutor`), which is fine IMO.

Reply:
@mposolda I see, I'll leave these lines you pointed out as they are. Only using the parsed result on `AuthzEndpointRequestObjectParser `.